core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+    fn get_value_mut_ref(&mut self, storage_pointer: KVStoragePointer) -> Option<&mut V> {
+        let entry = self.kv_pairs.get_mut(storage_pointer)?;
+        match entry {
+            InternalKVEntry::Occupied(occupied) => Some(occupied.value_ref_mut()),
+            InternalKVEntry::Vacant(_) => None,
+        }
+    }
+
+    /// Returns a reference to the key/value pair referenced by `handle`, if available.
+    fn get_kv(&self, handle: KVHandle) -> Option<KVRef<K, V>> {
+        let node = self.get_node(handle.node()).expect("node must exist");
+        let key_ptr = node.pair(handle.idx()).as_ref()?;
+        let pair = self.get_kv_ref(*key_ptr)?;
+        Some(KVRef::new(pair))
+    }
+
+    /// Returns the value referenced by `handle`, if available.
+    fn get_value(&self, handle: KVHandle) -> Option<&V> {
+        let node = self.get_node(handle.node()).expect("node must exist");
+        let ptr = node.pair(handle.idx()).as_ref()?;
+        self.get_kv_ref(*ptr).map(|pair| pair.value_ref())
+    }
+
+    /// If a parent node is set for the node referenced by `handle`, a handle to
+    /// this parent node will be returned.
+    fn ascend(&self, handle: NodeHandle) -> Option<KVHandle> {
+        let node = self
+            .get_node(handle)
+            .expect("node to ascend from must exist");
+
+        node.parent().map(|parent| {
+            let idx = node
+                .parent_idx()
+                .expect("if parent exists, parent_idx always exist as well; qed");
+            KVHandle::new(parent, idx)
+        })
+    }
+
+    /// Creates a root node with `key` and `val`.
+    ///
+    /// Returns a pointer to the storage of the inserted pair.
+    fn create_root(&mut self, key: K, val: V) -> KVStoragePointer {
+        debug_assert!(self.is_empty());
+        debug_assert!(self.root().is_none());
+
+        let pair = KVPair::<K, V>::new(key, val);
+        let ptr = self.put_pair(pair);
+
+        let mut node = Node::<K, V>::new();
+        node.set_pair(0, Some(ptr));
+        node.set_len(1);
+
+        let index = self.put(node);
+        self.header.len = 1;
+        self.header.root = Some(index);
+
+        ptr
+    }
+
+    /// Inserts `key` and `val` at `handle`.
+    ///
+    /// Returns a reference to the inserted value.
+    fn insert_kv(&mut self, handle: KVHandle, key: K, val: V) -> KVStoragePointer {
+        let mut ins_pair_ptr;
+        let mut ins_edge;
+        let out_storage_pointer;
+
+        let pair = KVPair::<K, V>::new(key, val);
+        let pair_ptr = self.put_pair(pair);
+
+        let mut cur_parent = match self.insert_into_node(handle, pair_ptr) {
+            (Fit(_), idx) => return idx,
+            (Split(left, pair_ptr, right), idx) => {
+                ins_pair_ptr = pair_ptr;
+                ins_edge = right;
+                out_storage_pointer = idx;
+                self.ascend(left.node())
+            }
+        };
+
+        loop {
+            match cur_parent {
```

Comment:
Please add a comment providing a rational that explains why this loop always breaks.